ftdc/custom_format_test.go

```diff
@@ -152,8 +153,116 @@ func TestCustomFormatRoundtripRich(t *testing.T) {
 }
 
 func TestReflection(t *testing.T) {
-	test.That(t, getFieldsForItem(&Basic{100}), test.ShouldResemble,
+	fields, err := getFieldsForStruct(reflect.TypeOf(&Basic{100}))
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, fields, test.ShouldResemble,
 		[]string{"Foo"})
-	test.That(t, getFieldsForItem(&Statser1{100, 0, 44.4}), test.ShouldResemble,
+
+	fields, err = getFieldsForStruct(reflect.TypeOf(Statser1{100, 0, 44.4}))
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, fields, test.ShouldResemble,
 		[]string{"Metric1", "Metric2", "Metric3"})
 }
+
+type TopLevel struct {
+	X      int
+	Nested Nested
+}
+
+type Nested struct {
+	Y      int
+	Deeper struct {
+		Z uint8
+	}
+}
+
+func TestNestedReflection(t *testing.T) {
+	val := &TopLevel{100, Nested{200, struct{ Z uint8 }{255}}}
+	fields, err := getFieldsForStruct(reflect.TypeOf(val))
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, fields, test.ShouldResemble,
+		[]string{"X", "Nested.Y", "Nested.Deeper.Z"})
+}
+
+type Complex struct {
+	F1 float32
+	F2 struct {
+		F3 float32
+		F4 float32
+	}
+	F5 struct {
+		F6 float32
+	}
+	F7  float32
+	F8  struct{}
+	F9  float32
+	F10 struct {
+		F11 float32
+		F12 float32
+		F13 float32
+	}
+	F14 struct {
+		F15 struct {
+			F16 struct {
+				F17 float32
+			}
+		}
+	}
+}
+
+func TestNestedReflectionParity(t *testing.T) {
+	// We use reflection in two paths:
+	// - Walking a "stats" object to create a schema.
+	// - Walking a "stats" object to get/flatten all of the values.
+	//
+	// It's critical that these two walks happen in the same order. Such that keys from the schema
+	// walk match their corresponding values.
+	complexObj := Complex{
+		F1: 1,
+		F2: struct {
+			F3 float32
+			F4 float32
+		}{3, 4},
+		F5: struct {
+			F6 float32
+		}{6},
+		F7: 7,
+		// F8 is tricky -- there are no leaf nodes. Therefore it must not be part of neither the
+		// schema nor value output.
+		F8: struct{}{},
+		F9: 9,
+		F10: struct {
+			F11 float32
+			F12 float32
+			F13 float32
+		}{11, 12, 13},
+		F14: struct {
+			F15 struct {
+				F16 struct {
+					F17 float32
+				}
+			}
+		}{
+			F15: struct {
+				F16 struct {
+					F17 float32
+				}
+			}{
+				F16: struct {
+					F17 float32
+				}{17},
+			},
+		},
```

Comment:
can we get another one of these with the `F14.F15.F16.F17`'s value equal to 0?

Reply:
I mean I'd prefer not to :stuck_out_tongue: 

I'm not 100% sure what your asking, nor what you think is risky. Are you asking for:
* Creating a second `Complex` object, just with a different value for F17
* Or amending the existing `Complex` object to have a `F18.F19.F20.F21: 0` value

Neither of those are particularly fun to write out, but I can. But maybe simply amending this `Complex` to include an `F18: 0` sibling of F17 is suitable?

Reply:
I meant:

> Or amending the existing Complex object to have a F18.F19.F20.F21: 0 value

It's fine. I realized that it would just be testing a more complex case of `F8`, which you already have and would probably not be adding much value. 

This is fine as is.